### PR TITLE
[FIX] barcodes_gs1_nomenclature: custom separators

### PR DESCRIPTION
**Pull request by svs-odoo**
Since https://github.com/odoo/odoo/pull/104287 the GS1 separator defined in the nomenclature is no more used in `BarcodeParser` `gs1_decompose_extanded` but it is used in `barcodeService` `cleanBarcode` instead.
The issue with that is `cleanBarcode` is not called in every case.
It will if the barcode is actually scanned but it won't if the barcode is decoded from a picture (using the device's camera) or triggered through an event directly (tahnks to the debug console.)
That means if we want to support custom separators, we have to clean the barcode in `parse_gs1_rule_pattern`.

How to reproduce:
- Activate `stock_barcode` and chose the GS1 nomenclature;
- Create a product and set a valid barcode (right checksum, eg.: 00000000012348);
- In the Barcode App, create a new receipt and scan a GS1 barcode for this product containing also a quantity and using a custom separator (`#` can be use aswell by default, eg.: 3008#0100000000012348 for 8x this product) with using the device camera or by trigger the event through the console:
`odoo.__DEBUG__.services['web.core'].bus.trigger('barcode_scanned', "3008#0100000000012348", $(".o_web_client")[0]);`
-> Issue: the barcode is not recognized.

With this commit, the parser will get the nomenclature's separator and will use it like before in `gs1_decompose_extanded`.

Enterprise PR: odoo/enterprise#47292

### CODE REVIEW

**Review by Julien00859 — APPROVED** [review]
review for that huge modification inside of `addons/barcodes_gs1_nomenclature/models/ir_http.py`

File: addons/barcodes_gs1_nomenclature/static/src/js/tests/barcode_parser_tests.js [hunk 1]
```diff
@@ -251,12 +251,11 @@ QUnit.module('Barcode GS1 Parser', {
 
     QUnit.test('Test Alternative GS1 Separator (fnc1)', async function (assert) {
         assert.expect(6);
-        const barcodeNomenclature = new BarcodeParser({'nomenclature_id': 2});
+        const nomenclature = this.data['barcode.nomenclature'].records[1];
+        nomenclature.rules = this.data['barcode.rule'].records;
+        let barcodeNomenclature = new BarcodeParser({ nomenclature_id: 2, nomenclature });
```
Comment (clesgow) [inline addons/barcodes_gs1_nomenclature/static/src/js/tests/barcode_parser_tests.js:256]:
```suggestion
        let barcodeNomenclature = new BarcodeParser({ nomenclature });
```
I'm not too sure about this, but the `load()` function will do two RPC calls to get the `nomenclature` we already gave based on the `nomenclature_id`. I guess we could spare those for this test.

**Review by clesgow — COMMENTED** [review]
Only have a few nitpicks on the test, but the fix itself looks good to me. :slightly_smiling_face:

File: addons/barcodes_gs1_nomenclature/static/src/js/tests/barcode_parser_tests.js (hunk 1 above)
Comment (svs-odoo) [reply addons/barcodes_gs1_nomenclature/static/src/js/tests/barcode_parser_tests.js:256]: Good idea, especially for QUnit tests where RPC should be mocked if we want to do things right

File: addons/barcodes_gs1_nomenclature/static/src/js/tests/barcode_parser_tests.js [hunk 2]
```diff
@@ -251,12 +251,11 @@ QUnit.module('Barcode GS1 Parser', {
 
     QUnit.test('Test Alternative GS1 Separator (fnc1)', async function (assert) {
         assert.expect(6);
-        const barcodeNomenclature = new BarcodeParser({'nomenclature_id': 2});
+        const nomenclature = this.data['barcode.nomenclature'].records[1];
```
[the remark this answers is not in the document]
Comment (svs-odoo) [reply addons/barcodes_gs1_nomenclature/static/src/js/tests/barcode_parser_tests.js:254]: I don't think so, the purpose of those tests are not to stick to demo data but to check chunk of codes do what is expected.
Python tests & tours are more suitable for that

**Pull request closed without merging**